File: code/search/src/binary_search/binary_search.rb
#   Part of Cosmos by OpenGenus Foundation

def binary_search(arr, l, r, x)
  return -1 if x.nil?

  if r >= l
    mid = (l + r) / 2
    if arr[mid] == x
      mid
    elsif arr[mid] > x
      binary_search(arr, l, mid - 1, x)
    else
      binary_search(arr, mid + 1, r, x)
    end
  end
end

arr = [3, 5, 12, 56, 92, 123, 156, 190, 201, 222]

number = 12

puts "Position of #{number} is #{binary_search(arr, 0, arr.length - 1, number)}"

File: code/search/src/binary_search/binary_search.hs
{-
    Part of Cosmos by OpenGenus Foundation
-}

binarySearch :: [Int] -> Int -> Int -> Int -> Int

binarySearch arr val low high
    | high < low = -1
    | arr!!mid < val = binarySearch arr val (mid+1) high
    | arr!!mid > val = binarySearch arr val low (mid-1)
    | otherwise = mid
    where
        mid = (low + high) `div` 2

main = do
    let arr = [3,5,12,56,92,123,156,190,201,222]

    let number = 12

    putStrLn("Position of "
             ++ show(number)
             ++ " is "
             ++ show(binarySearch arr number 0 ((length arr) - 1)))

